### PR TITLE
Mqtt

### PR DESCRIPTION
**Pull request by timthelion**

### CODE REVIEW

File: bspump/mqtt_metrics/__init__.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        self.dumper = JSONDumper(pretty=False)
+        self.container_id = None
+
+        self.sub_queue = []
+
+    def on_message(self, client, userdata, message):
+        payload = message.payload.decode("utf-8")
+        svc = self.App.get_service("bspump.PumpService")
+        if payload == "get":
```
Comment (timthelion) [inline bspump/mqtt_metrics/__init__.py:76]: We should match on:

```
message.topic
```

Rather than the payload

**Comment by timthelion** [comment]
When we were at the conference we met this company called quix and they build pipelines by connecting together docker containers. https://quix.io/

We'd like to parter with them, so that our pumps could run in their system. Afterall, our pumps are just docker containers so there is no reason they cannot be hooked into there system. Furthermore, I'd like to plan out this MQTT protocol with the idea that it can become not just an internal tool, but an industry standard for getting metrics from pipelines.

If you think about it, if quix is a system that strings together docker containers into pipelines, pumps are a set of pipelines (which can be connected via internal sources) and pipelines are a set of components. Then what we end up with is a hierarchy of pipelines:

```
Docker containers
 → Pumps
→→ Pipelines
→→→ Components
```

At each of these layers up to the component layer, we have something that can have
metrics about messages coming in and out
subcomponents
a topology showing how the data flows through the system

I'd like to make the directory structure such that it could support an infinitely deeply nested set of pipelines. Each component could have subcomponents ad infinity

I think this means that we change the path to look something like:

    Components/{container_id}/Components/{pipelien_id}/Components/{component_id}/events/subscribe

Why the repetition of "Components"? Because it allows for clean open namespacing.  Without such namespacing you end up with forbidden names, for example a component would not be allowed to be called Metrics

**Comment by timthelion** [comment]
My only question is if `Components` is too long a path element to be repeated 3 times ;)

Maybe

`kids/{container_id}/kids/{pipelien_id}/kids/{component_id}/events/subscribe`

Is better. Not sure. We can brain storm what would be best.

Or 

`_/{container_id}/_/{pipelien_id}/_/{component_id}/_/subscribe`

Nah

Hm,

```
metrics/get
events/subscribe
components/<component-name>/metrics/get
components/<component-name>/events/subscribe
components/<component-name>/components/<component-name>/metrics/get
components/<component-name>/components/<component-name>/components/<component-name>/metrics/get
```

Basically each directory has 3 subdirectories:

```
events
metrics
components
```

And maybe lets rename metrics to `topology` so it becomes:


```
events
topology
components
```

But components is still too long to be repeated so maybe

```
events
topology
kids
```

**Comment by timthelion** [comment]
@lukasvecerka23 do you have an opinion on the naming convention?

**Comment by timthelion** [comment]
ChatGPT is strongly against `kids` but likes `comps`. Or we could just go with `c`

**Comment by timthelion** [comment]
I'm liking c, here are some examples GPT made up:

```c/SolarArray/c/PhotonCollector/events/subscribe
c/PowerGrid/c/FlowController/topology/get
c/BatteryBank/c/ChargeMeter/events/subscribe
c/WindFarm/c/TurbulenceAnalyzer/topology/get
c/GenerationUnit/c/EnergyConverter/events/subscribe
c/DistributionLine/c/LoadBalancer/topology/get
c/StorageUnit/c/CapacityMonitor/events/subscribe
c/TransmissionLine/c/VoltageRegulator/topology/get
c/SolarPanel/c/IrradianceSensor/events/subscribe
c/EnergyStorage/c/DischargeController/topology/get
```

**Comment by lukasvecerka23** [comment]
I think `comps` or `c` is good. Better than kids, I'll stay with ChatGPT in this :D

**Comment by lukasvecerka23** [comment]
So in the mqtt_service in the `on_message` callback. There will be few conditions to match each pattern right? And on different topic level there will be different logic

**Comment by timthelion** [comment]
yes

**Comment by lukasvecerka23** [comment]
So there wouldn't be needed to use the configuration service? And are TeskaLabs okay tak we have our own fork of their ASAB publicly on github? 😅

**Comment by timthelion** [comment]
> So there wouldn't be needed to use the configuration service? And are TeskaLabs okay tak we have our own fork of their ASAB publicly on github? 😅

@lukasvecerka23 They don't have to be OK, it's open source and we also have acontract with them saying we can use it. We still need to use the configuration service, I just moved that to asab to get influx metrics working when net=host too.

**Pull request merged**